src/v/cluster/scheduling/constraints.cc

```diff
@@ -150,4 +152,38 @@ soft_constraint_evaluator least_allocated() {
     return soft_constraint_evaluator(std::make_unique<impl>());
 }
 
+soft_constraint_evaluator distinct_rack(
+  const std::vector<model::broker_shard>& replicas,
+  ss::noncopyable_function<ss::sstring(model::node_id)> get_rack) {
+    class impl : public soft_constraint_evaluator::impl {
+    public:
+        impl(
+          const std::vector<model::broker_shard>& replicas,
+          ss::noncopyable_function<ss::sstring(model::node_id)> gr)
+          : _replicas(replicas)
+          , _get_rack(std::move(gr)) {}
+        uint64_t score(const allocation_node& node) const final {
+            // if node.rack() should be different from others
+            // otherwise return 0
+            for (auto [node_id, shard] : _replicas) {
+                auto rack = _get_rack(node_id);
```

Comment:
would it be possible to get the optional here and return `1` if it is empty ? this way we would not affect scoring when there is no rack set.